internal/iteratoriterator.hpp

```diff
@@ -80,11 +80,11 @@ namespace iter {
         return ret;
       }
 
-      auto operator*() -> decltype(**sub_iter) {
+      auto operator*() const -> decltype(**sub_iter) {
```

Comment:
This places a greater requirement on the sub_iter to have a const operator*()

Reply:
If sub_iter is supposed to be an InputIterator, then it must have a const version of operator*. In C++17 standard:
[iterator.requirements.general]¶13 "In the following sections, a and b denote values of type X or const X"; then Table 95 "Input iterator requirements [...]" contains expression "*a".
Thus operator* must work on a const X.

If you think IteratorIterator should support also iterators that do not obey to the InputIterator requirements, both a const and non const version of the operators could be defined.

Reply:
While trying to add a test case for non-const operator* (i.e. we should not add code unless there is a test that requires that code), I realized that IteratorIterator requires that TopIter is a RandomAccessIterator (requires the random_access_iterator_tag; there is a static_assert for that). Thus unless a class lies about being such an iterator, it must support const operator*.  (Note that IteratorIterator also requires a const operator[], which I forgot in that version; I will correct that.)

Simple duplication of current operators *, -> and [], with const and non-const versions, would support an iterator having both the const and non-const versions (and also a standard iterator having only the const version).  To do the tests requires to define a new special iterator, as vector::iterator does not have non-const versions.

To support non standard conforming iterators, without the const version, would require to remove the decltype return type of the operators, or use SFINAE, on the const version in IteratorIterator.  Is this what you would like?

Reply:
Ok, so I added tests to verify that IteratorIterator operators are all working as specified by RandomAccessIterator in the standard, for both const and non-const iterators (having only const versions of *, -> and []).  I passed each line of the requirements tables, I hope I did not miss anything (but I only test method results, I do not test the types).

I corrected the operator[] to be also const (as mentioned in my previous comment), to pass the RandomAccessIterator tests.

Those tests showed me that a weird operator- was not covered; in fact it was an incorrect operator where an expression like (2 - it) (subtracting an iterator from an integer) would give (it - 2).  I thus simply removed that method.